From 1e9039b6d853c834b07adeee0aa82ca73794d673 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Wed, 4 Oct 2017 17:56:57 +0200
Subject: [PATCH 02/41] Disable echo

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 4215787..17cd997 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -982,6 +982,7 @@ RokiSimulatorItemImpl::RokiSimulatorItemImpl(RokiSimulatorItem* self)
     contactFileName.clear();
     zkcFileName.clear();
     //useWorldCollision = false;
+    zEchoOff();
 
 }
 
-- 
2.7.4

