From 17ef0ae958f502638be6a0d9d4fa90052b0d19c2 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Thu, 5 Oct 2017 19:48:42 +0200
Subject: [PATCH 05/41] Fix name of optical info

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 20 +++++++++-----------
 1 file changed, 9 insertions(+), 11 deletions(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 2e9b2c8..ad1a6ae 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -1003,23 +1003,21 @@ bool RokiOpticalInfo::operator!=(const RokiOpticalInfo& other) const
 
 string RokiOpticalInfo::createName(Vector3f amb, Vector3f dif, Vector3f spc, float ns, float sns, float alpha)
 {
-    string name = "#";
-    stringstream stream;
+    char name[256];
 
     amb *= 255.0f;
     dif *= 255.0f;
     spc *= 255.0f;
     ns  *= 255.0f;
-    sns *= 255.0f;
+    // sns *= 255.0f;
     alpha *= 255.0f;
-    stream << std::hex
-           << (int)amb[0] << (int)amb[1] << (int)amb[2]
-           << (int)dif[0] << (int)dif[1] << (int)dif[2]
-           << (int)spc[0] << (int)spc[1] << (int)spc[2]
-           << (int)ns << (int)sns << (int)alpha;
-    name += stream.str();
-
-    return name;
+    sprintf(name, "#%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x",
+            (int)amb[0], (int)amb[1], (int)amb[2],
+            (int)dif[0], (int)dif[1], (int)dif[2],
+            (int)spc[0], (int)spc[1], (int)spc[2],
+            (int)ns, (int)sns, (int)alpha);
+
+    return string(name);
 }
 
 
-- 
2.7.4

