From 8022fba6c32706807b0c1f6326c869c5754b63c2 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Fri, 6 Oct 2017 18:04:21 +0200
Subject: [PATCH 06/41] Add class to contain multiple optical infos

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index ad1a6ae..f6f8474 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -227,6 +227,19 @@ private:
     string createName(Vector3f amb, Vector3f dif, Vector3f spc, float ns, float sns, float alpha);
 };
 
+
+class RokiOpticalInfoList
+{
+public:
+    bool add(RokiOpticalInfo opt);
+    RokiOpticalInfo& get(int index) { return list_[index]; }
+    const RokiOpticalInfo& get(int index) const { return list_[index]; }
+    int num() const { return list_.size(); }
+
+private:
+    vector<RokiOpticalInfo> list_;
+};
+
 }
 
 namespace cnoid {
@@ -1040,6 +1053,17 @@ void RokiOpticalInfo::exportToZOpticalInfo(zOpticalInfo* opt)
 }
 
 
+bool RokiOpticalInfoList::add(RokiOpticalInfo opt)
+{
+    for(auto o : list_){
+        if(o == opt) return false;
+    }
+    list_.push_back(opt);
+
+    return true;
+}
+
+
 void RokiSimulatorItem::initializeClass(ExtensionManager* ext)
 {
     ext->itemManager().registerClass<RokiSimulatorItem>(N_("RokiSimulatorItem"));
-- 
2.7.4

