From ae7b1ef3f7970c963670a907f98939e11f1b3e17 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Mon, 9 Oct 2017 17:46:59 +0200
Subject: [PATCH 12/41] Fix to return original name if it exists

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 3e48e31..14ea746 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -356,6 +356,9 @@ string RokiOpticalInfo::createName()
     float alpha = alpha_ * 255.0f;
     char name[256];
 
+    if(!SgObject::name().empty())
+        return SgObject::name();
+
     sprintf(name, "#%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x",
             (int)amb[0], (int)amb[1], (int)amb[2],
             (int)dif[0], (int)dif[1], (int)dif[2],
-- 
2.7.4

