From 5a141d9f8b0e6bbe0eff7b16df8506844be0c4ae Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Tue, 10 Oct 2017 16:39:37 +0200
Subject: [PATCH 14/41] Fix shape name

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 724c48c..9753086 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -219,6 +219,9 @@ public:
     void getKinematicStateFromRoki();
     void setKinematicStateToRoki(zVec dis, int k);
     void setTorqueToRoki();
+
+private:
+    string createShapeName();
 };
 typedef ref_ptr<RokiLink> RokiLinkPtr;
 
@@ -710,6 +713,15 @@ void RokiLink::createLink(RokiSimulatorItemImpl* simImpl, RokiBody* body, const
 }
 
 
+string RokiLink::createShapeName()
+{
+    stringstream stream;
+
+    stream << link->name() << "#" << zListNum(rkLinkShapeList(rklink));
+    return stream.str();
+}
+
+
 void RokiLink::createGeometry()
 {
     if(link->collisionShape()){
@@ -719,7 +731,7 @@ void RokiLink::createGeometry()
                 zShape3D* sp = zAlloc( zShape3D, 1 );
                 zShape3DInit(sp);
                 zShape3DType(sp) = zShapeTypeByStr((char*)"polyhedron");
-                zNameSet( sp, const_cast<char*>(link->name().c_str()) );
+                zNameSet( sp, const_cast<char*>(createShapeName().c_str()) );
                 sp->com = &zprim_ph3d_com;
 
                 zPH3D* ph = (zPH3D*)&sp->body;
@@ -852,7 +864,7 @@ void RokiLink::addMesh(MeshExtractor* extractor)
             if(created){
                 shapes.push_back(sp);
                 zBox3DInit( zShape3DBB(sp) );
-                zNameSet( sp, const_cast<char*>(link->name().c_str()) );
+                zNameSet( sp, const_cast<char*>(createShapeName().c_str()) );
                 rkLinkShapePush( rklink, sp );
                 meshAdded = true;
             }
-- 
2.7.4

