From 59c9ec254c7458c6dfb8bea2f475455ea4707640 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Tue, 10 Oct 2017 17:09:47 +0200
Subject: [PATCH 16/41] Remove continue

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index ff5fbfe..4a8b47b 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -1184,7 +1184,6 @@ void RokiBody::findAllOpticalInfo(RokiOpticalInfoList& list)
         if(!(group = dynamic_cast<SgGroup*>(link->visualShape()))){
             cerr << "cannot be converted to 'SgGroup*': "
                  << typeid(link->visualShape()).name() << endl;
-            continue;
         }
         for(auto child : *group){
             list.add(RokiOpticalInfo(child.get()));
-- 
2.7.4

