From 476953e8f1f5b38e88227881ca502793c0ba24e4 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Tue, 10 Oct 2017 18:56:07 +0200
Subject: [PATCH 17/41] Fix to find optic name when adding shape to RokiLink

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 4a8b47b..5d14587 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -190,8 +190,10 @@ private:
 class RokiShapeList
 {
 public:
-    bool add(zShape3D* shape);
+    bool add(zShape3D* shape, string opticName);
     zShape3D* get(int index) { return list_[index]; }
+    string getOpticName(zShape3D* shape) { return opticNameMap_[shape]; }
+    string getOpticName(int index) { return getOpticName(get(index)); }
     int num() const { return list_.size(); }
 
     vector<zShape3D*>::iterator begin() { return list_.begin(); }
@@ -201,6 +203,7 @@ public:
 
 private:
     vector<zShape3D*> list_;
+    map<zShape3D*, string> opticNameMap_;
 };
 
 
@@ -215,6 +218,7 @@ public:
     Link* link;
     rkLink* rklink;
     vector<zShape3D*> shapes;
+    map<zShape3D*, string> opticNameMap;
     vector<Vector3> vertices;
     vector<Triangle> triangles;
     Matrix3 wAtt;  // world
@@ -476,9 +480,10 @@ bool RokiOpticalInfoList::add(RokiOpticalInfo opt)
 }
 
 
-bool RokiShapeList::add(zShape3D* shape)
+bool RokiShapeList::add(zShape3D* shape, string opticName)
 {
     list_.push_back(shape);
+    opticNameMap_[shape] = opticName;
     return true;
 }
 
@@ -779,6 +784,9 @@ void RokiLink::createGeometry()
                 zAABox3DToBox3D( &aabb, zShape3DBB(sp) );
                 rkLinkShapePush( rklink, sp );
             }
+            zShapeListCell *cp = zListTail(rkLinkShapeList(rklink));
+            RokiOpticalInfo opt(link->visualShape());
+            opticNameMap[cp->data] = opt.name();
         }
         delete extractor;
     }
-- 
2.7.4

