From 29a874e82987aca9387e49c1f245bc56c2962f51 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Wed, 11 Oct 2017 10:37:23 +0200
Subject: [PATCH 19/41] Fix minor

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index ff60b65..b87c135 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -1243,7 +1243,8 @@ void RokiBody::exportAsZkcFile(const std::string& filename)
         zOpticalInfo *oarray = rkChainShape(chain)->optic.buf;
         zOpticalInfo *opt;
         string optName = shapeList.getOpticName(shapeList.get(i));
-        zNameFind(oarray, opticList.num(), const_cast<char*>(optName.c_str(), opt);
+        zNameFind(oarray, opticList.num(),
+                  const_cast<char*>(optName.c_str()), opt);
         zShape3DClone(shapeList.get(i), target, opt);
     }
 
-- 
2.7.4

