From 46cfc1d10e9408640b8ca6d938d0320edb4f0037 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Wed, 11 Oct 2017 10:37:29 +0200
Subject: [PATCH 20/41] Fix how to find RokiLink

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index b87c135..9017056 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -1204,7 +1204,7 @@ void RokiBody::findAllOpticalInfo(RokiOpticalInfoList& list)
 void RokiBody::findAllShape(RokiShapeList& list)
 {
     for(int i=0; i<body()->numLinks(); ++i){
-        RokiLink *link = dynamic_cast<RokiLink*>(body()->link(i));
+        RokiLink *link = rokiLinkMap[body()->link(i)];
         if(link){
             zShapeListCell *cp;
             zListForEach(rkLinkShapeList(link->rklink), cp){
-- 
2.7.4

