From 5dd250764aa824204596761da1e9464d757e1c1d Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Wed, 11 Oct 2017 11:25:07 +0200
Subject: [PATCH 21/41] Remove unused local variable

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 9017056..0777e01 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -398,7 +398,6 @@ void RokiOpticalInfo::visitShape(SgShape* shape)
     if(material){
         scaleToRokiCompatible(material);
     }
-    SgMesh* mesh = shape->mesh();
 }
 
 
-- 
2.7.4

