From 8c59a1cb9b47d7f8b48d33ff7d9aa3af7792d479 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Wed, 11 Oct 2017 21:24:30 +0200
Subject: [PATCH 27/41] Add switch between collision and visual shapes

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 8dde073..243f75f 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -312,6 +312,8 @@ public:
     void setTorqueToRoki();
     void updateForceSensors();
 
+    void switchToColShape();
+    void switchToVisShape();
     void findAllOpticalInfo(RokiOpticalInfoList& list);
     void findAllShape(RokiVisualShapeList& list);
     void exportAsZkcFile(const std::string& filename);
@@ -1494,6 +1496,26 @@ void RokiBody::updateForceSensors()
 }
 
 
+void RokiBody::switchToColShape()
+{
+    for(int i=0; i<body()->numLinks(); ++i){
+        Link* link = body()->link(i);
+        RokiLink *rklink = rokiLinkMap[link];
+        rklink->switchToColShape();
+    }
+}
+
+
+void RokiBody::switchToVisShape()
+{
+    for(int i=0; i<body()->numLinks(); ++i){
+        Link* link = body()->link(i);
+        RokiLink *rklink = rokiLinkMap[link];
+        rklink->switchToVisShape();
+    }
+}
+
+
 void RokiBody::findAllOpticalInfo(RokiOpticalInfoList& list)
 {
     for(int i=0; i<body()->numLinks(); ++i){
@@ -1514,7 +1536,7 @@ void RokiBody::findAllShape(RokiVisualShapeList& list)
 {
     for(int i=0; i<body()->numLinks(); ++i){
         Link* link = body()->link(i);
-        RokiLink *rklink = rokiLinkMap[body()->link(i)];
+        RokiLink *rklink = rokiLinkMap[link];
         SgGroup *group;
         if(!(group = dynamic_cast<SgGroup*>(link->visualShape()))){
             cerr << "cannot be converted to 'SgGroup*': "
@@ -1546,6 +1568,7 @@ void RokiBody::exportAsZkcFile(const std::string& filename)
 
     // find all shapes
     RokiVisualShapeList shapeList;
+    switchToVisShape();
     findAllShape(shapeList);
     zArrayAlloc(&rkChainShape(chain)->shape, zShape3D, shapeList.num());
     for(int i=0; i<shapeList.num(); ++i){
@@ -1558,6 +1581,7 @@ void RokiBody::exportAsZkcFile(const std::string& filename)
 
     // export
     rkChainWriteFile(chain, (char *)filename.c_str());
+    switchToColShape();
 }
 
 
-- 
2.7.4

