From ecc2eb5996c0d8c36f6ae37b1a6490020108e097 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Thu, 12 Oct 2017 12:35:32 +0200
Subject: [PATCH 29/41] Fix to use functions to update shape list

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 0ea403f..f26ccb2 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -817,16 +817,14 @@ RokiLink::RokiLink
     createLink(simImpl, rokiBody, o, stuffisLinkName);
     rkChainMass(rokiBody->chain) += link->mass();
 
+    updateVisShapeList();
     createGeometry();
+    updateColShapeList();
 
     for(Link* child = link->child(); child; child = child->sibling()){
         new RokiLink(simImpl, rokiBody, this, o, child, stuffisLinkName);
     }
 
-    zListInit(&colShapeList);
-    zListInit(&visShapeList);
-    overwriteListRoot(&colShapeList, rkLinkShapeList(rklink));
-
 }
 
 
@@ -1577,7 +1575,8 @@ void RokiBody::findAllOpticalInfo(RokiOpticalInfoList& list)
                  << typeid(link->visualShape()).name() << endl;
         }
         for(auto child : *group){
-            list.add(RokiOpticalInfo(child.get()));
+            RokiOpticalInfo *opt = new RokiOpticalInfo(child.get());
+            list.add(RokiOpticalInfo(*opt));
         }
     }
 }
@@ -1594,9 +1593,10 @@ void RokiBody::findAllShape(RokiVisualShapeList& list)
                  << typeid(link->visualShape()).name() << endl;
         }
         for(auto child : *group){
-            RokiVisualShape shape(child.get(), *rklink);
-            list.add(shape);
+            RokiVisualShape *shape = new RokiVisualShape(child.get(), *rklink);
+            list.add(*shape);
         }
+        rklink->updateVisShapeList();
     }
 }
 
@@ -1632,7 +1632,7 @@ void RokiBody::exportAsZkcFile(const std::string& filename)
 
     // export
     rkChainWriteFile(chain, (char *)filename.c_str());
-    switchToColShape();
+    // switchToColShape();
 }
 
 
-- 
2.7.4

