From d9c3e4af53b2057b21c26ef0defff0c195818d81 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Thu, 12 Oct 2017 12:36:31 +0200
Subject: [PATCH 30/41] Remove code for debug

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 23 -----------------------
 1 file changed, 23 deletions(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index f26ccb2..e1388ed 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -1336,32 +1336,9 @@ void RokiLink::updateListSet(ListSet *listSet, zShapeList *list)
 
 void RokiLink::overwriteList(zShapeList *list, ListSet *listSet)
 {
-    zShapeListCell *cp;
-    cout << "==============" << endl;
-    cout << "number = " << zListNum(list) << endl;
-    cout << "root: " <<zListRoot(list) << endl;
-    for(int i=0; i<zListNum(list); ++i){
-        zListItem(list, i, &cp);
-        cout << "<" << i << ">";
-        cout << "cell [" << cp << "] ";
-        cout << zListCellPrev(cp) << " <- prev | next-> "
-             << zListCellNext(cp) << endl;
-    }
-
     zListSetNum(list, listSet->num);
     zListCellSetPrev(zListRoot(list), listSet->head);
     zListCellSetNext(zListRoot(list), listSet->tail);
-
-    cout << "---------------" << endl;
-    cout << "number = " << zListNum(list) << endl;
-    cout << "root: " <<zListRoot(list) << endl;
-    for(int i=0; i<zListNum(list); ++i){
-        zListItem(list, i, &cp);
-        cout << "<" << i << ">";
-        cout << "cell [" << cp << "] ";
-        cout << zListCellPrev(cp) << " <- prev | next-> "
-             << zListCellNext(cp) << endl;
-    }
 }
 
 
-- 
2.7.4

