From 272b75c60d3bc48efcae7e9e6bb94c59adbb0b17 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Thu, 12 Oct 2017 13:32:34 +0200
Subject: [PATCH 32/41] Add brackets

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 2d40299..d4a5465 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -622,28 +622,33 @@ string RokiVisualShape::createShapeName()
 bool RokiVisualShape::canBeCreatedAsPrimitive(SgMesh* mesh)
 {
     const int type = mesh->primitiveType();
-    if(type == SgMesh::MESH)
+    if(type == SgMesh::MESH){
         return false;
+    }
 
-    if(!isCurrentScaled)
+    if(!isCurrentScaled){
         return true;
+    }
 
     Affine3 S = currentTransformWithoutScaling.inverse() *\
         currentTransform;
-    if(!S.linear().isDiagonal())
+    if(!S.linear().isDiagonal()){
         return false;
+    }
 
     Vector3 scale = S.linear().diagonal();
     if(type == SgMesh::BOX){
         return true;
     } else if(type == SgMesh::SPHERE){
         // check if the sphere is uniformly scaled for all the axes
-        if(scale.x() == scale.y() && scale.x() == scale.z())
+        if(scale.x() == scale.y() && scale.x() == scale.z()){
             return true;
+        }
     } else if(type == SgMesh::CYLINDER || type == SgMesh::CONE){
         // check if the bottom circle face is uniformly scaled
-        if(scale.x() == scale.z())
+        if(scale.x() == scale.z()){
             return true;
+        }
     }
     return false;
 }
-- 
2.7.4

