From fd27ea44f131ac637a02855d8aea027de5948110 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Thu, 12 Oct 2017 14:43:50 +0200
Subject: [PATCH 35/41] Fix direction of cylinder and cone

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index b6bcef9..5994dcf 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -695,10 +695,10 @@ void RokiVisualShape::createAsPrimitive(SgMesh* mesh)
         break; }
     case SgMesh::CYLINDER : {
         SgMesh::Cylinder cylinder = mesh->primitive<SgMesh::Cylinder>();
-        Vector3 ay(T_(0,1), T_(1,1), T_(2,1));
-        ay *= cylinder.height/2.0 * scale.y();
-        Vector3 c1_ = T0 * ay.cast<Position::Scalar>();
-        Vector3 c2_ = T0 * (-ay).cast<Position::Scalar>();
+        Vector3 ax(T_(0,0), T_(1,0), T_(2,0));
+        ax *= cylinder.height/2.0 * scale.x();
+        Vector3 c1_ = T0 * ax.cast<Position::Scalar>();
+        Vector3 c2_ = T0 * (-ax).cast<Position::Scalar>();
         zVec3D c1, c2;
         zVec3DCreate(&c1, c1_.x(), c1_.y(), c1_.z());
         zVec3DCreate(&c2, c2_.x(), c2_.y(), c2_.z());
@@ -706,10 +706,10 @@ void RokiVisualShape::createAsPrimitive(SgMesh* mesh)
         break; }
     case SgMesh::CONE : {
         SgMesh::Cone cone = mesh->primitive<SgMesh::Cone>();
-        Vector3 ay(T_(0,1), T_(1,1), T_(2,1));
-        ay *= cone.height/2.0 * scale.y();
-        Vector3 c1_ = T0 * ay.cast<Position::Scalar>();
-        Vector3 c2_ = T0 * (-ay).cast<Position::Scalar>();
+        Vector3 ax(T_(0,0), T_(1,0), T_(2,0));
+        ax *= cone.height/2.0 * scale.x();
+        Vector3 c1_ = T0 * ax.cast<Position::Scalar>();
+        Vector3 c2_ = T0 * (-ax).cast<Position::Scalar>();
         zVec3D c1, c2;
         zVec3DCreate(&c1, c1_.x(), c1_.y(), c1_.z());
         zVec3DCreate(&c2, c2_.x(), c2_.y(), c2_.z());
-- 
2.7.4

