From bc65d4d505a981751b5e72b1c8e2007a0aff1fd8 Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Thu, 12 Oct 2017 15:08:44 +0200
Subject: [PATCH 36/41] Remove opticNameMap

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 5994dcf..67ad963 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -256,7 +256,6 @@ public:
     Link* link;
     rkLink* rklink;
     vector<zShape3D*> shapes;
-    map<zShape3D*, string> opticNameMap; // deprecated
     vector<Vector3> vertices;
     vector<Triangle> triangles;
     Matrix3 wAtt;  // world
@@ -347,6 +346,7 @@ private:
     RokiBody* rokiBody;
 };
 
+
 }
 
 namespace cnoid {
@@ -729,7 +729,7 @@ void RokiVisualShape::createAsPolyhedron(SgMesh* mesh)
 {
     vector<Vector3> vertices;
     vector<Triangle> triangles;
-    
+
     Affine3 T = currentTransform;
     Affine3 invT = Affine3::Identity();
     invT.linear() = link_.wAtt.transpose();
@@ -1089,9 +1089,6 @@ void RokiLink::createGeometry()
                 zAABox3DToBox3D( &aabb, zShape3DBB(sp) );
                 rkLinkShapePush( rklink, sp );
             }
-            zShapeListCell *cp = zListTail(rkLinkShapeList(rklink));
-            RokiOpticalInfo opt(link->visualShape());
-            opticNameMap[cp->data] = opt.name();
         }
         delete extractor;
     }
-- 
2.7.4

