From a02e165971b9faa3ed74a03114105bedcf23c7fe Mon Sep 17 00:00:00 2001
From: Hiroshi Atsuta <atsuta.hiroshi@gmail.com>
Date: Thu, 12 Oct 2017 15:24:13 +0200
Subject: [PATCH 37/41] Fix redundant code

---
 src/RokiPlugin/RokiSimulatorItem.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/RokiPlugin/RokiSimulatorItem.cpp b/src/RokiPlugin/RokiSimulatorItem.cpp
index 67ad963..8e69af9 100644
--- a/src/RokiPlugin/RokiSimulatorItem.cpp
+++ b/src/RokiPlugin/RokiSimulatorItem.cpp
@@ -1557,7 +1557,7 @@ void RokiBody::findAllOpticalInfo(RokiOpticalInfoList& list)
         }
         for(auto child : *group){
             RokiOpticalInfo *opt = new RokiOpticalInfo(child.get());
-            list.add(RokiOpticalInfo(*opt));
+            list.add(*opt);
         }
     }
 }
-- 
2.7.4

